PLU-428 [FIND-MULTIPLE-ROWS-1]: Tiles

kevinkim-ogp opened the pull request:
### TL;DR

Added a new "Find Multiple Rows" action to the Tiles app that allows users to retrieve multiple rows from a tile based on specified filters.

This is a prerequisite for for-each.

### What changed?

- Created a new `findMultipleRows` action that retrieves multiple rows from a tile based on filter conditions
- Implemented permission checks to ensure only owners and editors can access the data
- Added support for sorting results in ascending or descending order
- Implemented scan limit functionality to control the number of rows returned
- Created comprehensive tests to verify the action's functionality
- Added appropriate error handling for invalid filters and deleted tiles
- Integrated the new action with the existing Tiles app

### How to test?

1. Create a new action and select the "Find multiple rows" event from the Tiles app
2. Select a tile and configure filter conditions to match the rows you want to retrieve
3. Choose the order of results (ascending or descending)
4. Test the step and verify that the correct rows are returned


### Note to reviewers
This is meant to be the backend implementation of the findMultipleRows action, the data-out metadata will be updated again.

Feature flag has been added on LaunchDarkly to allow for progressive rollout.

kevinkim-ogp commented:
* **#1026** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/1026?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2  dependent PRs ([#1029](https://github.com/opengovsg/plumber/pull/1029) <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/1029?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#1065](https://github.com/opengovsg/plumber/pull/1065) <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/1065?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#1025** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/1025?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1023** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/1023?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#923** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/923?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#900** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/900?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#899** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/899?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/899?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `trunk/for-each`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

m0nggh commented on packages/backend/src/apps/tiles/actions/find-multiple-rows/index.ts:
```diff
@@ -0,0 +1,250 @@
+          columns: columnData,
+        },
+        columns: consolidatedColumns,
+        rowId: 'rowId',
```
can remove `rowId: rowId` because it seems like you won' need it

m0nggh commented on packages/backend/src/apps/tiles/actions/find-multiple-rows/get-data-out-metadata.ts:
```diff
@@ -0,0 +1,60 @@
+      label: 'Number of rows found',
+      order: 2,
+    },
+    rowId: {
+      label: 'Row ID',
+      value: 'rowId',
+      displayedValue: ' ',
+    },
```
this can be removed too

m0nggh reviewed (approved):
tested that no rows found works, multiple rows can be retrieved both asc and desc order

checked that checkbox field for formsg still works

checked that LD flag is set properly, lgtm!

m0nggh commented on packages/backend/src/apps/tiles/actions/find-multiple-rows/get-data-out-metadata.ts:
```diff
@@ -0,0 +1,55 @@
+    },
+  }
+
+  const columnMetadata: IDataOutMetadata = {}
+  if (dataOut.columns) {
+    Object.entries(dataOut.columns).forEach(([name, { id }]) => {
+      columnMetadata[name] = {
+        id: { isHidden: true },
+        value: {
+          label: name,
+          order: columnOrder[id] ? columnOrder[id] + 2 : null,
+        },
+      }
+    })
+  }
```
ian: should use id as the key instead

kevinkim-ogp commented:
@m0nggh thanks for the intial review, will need your help to review again 🙏 

made the following changes based on comments from ian and yourself:
- refactored `columns` output to an array to preserve order (instead of adding using an object with the column name as key, which could cause errors when columns have the same name)
- added `schema.ts` to validate the dataOut

m0nggh commented:
okok thks for the quick fix, tested with same column names and it works, gogo
